### PR TITLE
 curvefs/client: fix datacache flush get inode failed cause core dump

### PR DESCRIPTION
**Pull request by baijiaruo**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 1]
```diff
@@ -1012,6 +1012,24 @@ void ChunkCacheManager::ReleaseCache(S3ClientAdaptorImpl* s3ClientAdaptor) {
...
+        dataWCacheMap_.erase(key);
+        s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(1);
+        LOG(INFO) << "chunk flush DataCacheByteDec1 len:"
+                  << dataCacheLen;
+        s3ClientAdaptor_->GetFsCacheManager()->
+            DataCacheByteDec(dataCacheLen);
+        if (!s3ClientAdaptor_->GetFsCacheManager()
+            ->WriteCacheIsFull()) {
+            LOG(INFO) << "write cache is not full, signal wait.";
+            s3ClientAdaptor_->GetFsCacheManager()->FlushSignal();
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1028]: it seems these code are not necessary protected by lock

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 2]
```diff
@@ -1227,15 +1234,14 @@ CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
...
-            dirty_.exchange(false, std::memory_order_acq_rel);
             ret = s3ClientAdaptor_->AllocS3ChunkId(fsId, &chunkId);
             if (ret != FSStatusCode::OK) {
                 LOG(ERROR) << "alloc s3 chunkid fail. ret:" << ret;
-                delete data;
                 return CURVEFS_ERROR::INTERNAL;
             }
+            data = new char[len_];
+            memcpy(data, data_, len_);
+            dirty_.exchange(false, std::memory_order_acq_rel);
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1244]: dirty_.store is enough

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 3]
```diff
@@ -1259,6 +1265,8 @@ CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
             }
             if (ret < 0) {
                 LOG(ERROR) << "write object fail. object: " << objectName;
+                delete data;
+                dirty_.exchange(true, std::memory_order_acq_rel);
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1269]: ditto

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 4]
```diff
@@ -1277,6 +1285,7 @@ CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
                     inodeId, inodeWrapper);
             if (ret != CURVEFS_ERROR::OK) {
                 LOG(ERROR) << "get inode fail, ret:" << ret;
+                dirty_.exchange(true, std::memory_order_acq_rel);
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:1288]: ditto

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 2 above)
Comment (wu-hanqing) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1244]: `len_` in line 1219, and `createTime_` in line 1221 are not within lock

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 1 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:1028]: fixed

**Pull request merged**